### PR TITLE
fix #5957 feat(nimbus-ui): sortable directory columns

### LINKED ISSUE #5957: Make directory view columns sortable
Let's make the directory view table sortable by each column header.  Looks like there's a [react-bootstrap-table](https://github.com/react-bootstrap-table/react-bootstrap-table2) package that should do this, [example in storybook](https://react-bootstrap-table.github.io/react-bootstrap-table2/storybook/index.html?selectedKind=Sort%20Table&selectedStory=Default%20Sort%20Table&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel)

We only need to be able to sort by a single column at a time.  Currently applied sorting should be reflected in the url a query parameter.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1493)

### PR DESCRIPTION
**Pull request by lmorchard**
Because:

* We want to make the directory view table sortable by each column

* Currently applied sorting should be reflected in the url query
  parameter.

This commit:

* Adds support for directory tab selection via query parameter

* Adds experiment sorting to DirectoryTable based on location query
  parameters

* Adds SortableColumnTitle component to DirectoryTable to cycle through
  column sorting modes and update location query parameters

* Updates column definitions in DirectoryTable components to support
  sorting criteria

* Adds sorting utilities to lib/experiment.ts

* Tweaks mock data to add more sortable variety

* Updates tests and stories with router context

* Adds useSearchParamsState hook to read and update location query
  parameters

### CODE REVIEW

**Comment by lmorchard** [comment]
Some initially interesting stories:

* https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/2d9eb7a54dd43fb2e6bc6d75c29d07923efb4a09/nimbus-ui/index.html?path=/story/pages-home-directorytable--basic
* https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/2d9eb7a54dd43fb2e6bc6d75c29d07923efb4a09/nimbus-ui/index.html?path=/story/pages-home--basic

**Comment by jaredlockhart** [comment]
HOLY COW @lmorchard IT'S SO COOL 👏 👏 👏 👏  🎉 🎉 🎉 🎉 🎉 🎉

**Comment by lmorchard** [comment]
Updated stories (should basically look the same):

* https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/5d25299042304362b0f0f0006478ba8d780688d4/nimbus-ui/index.html?path=/story/pages-home-directorytable--basic
* https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/5d25299042304362b0f0f0006478ba8d780688d4/nimbus-ui/index.html?path=/story/pages-home--basic

**Comment by lmorchard** [comment]
Hmm, almost passing CI. I see an integration_nimbus failure, but not sure if it's intermittent (seems not entirely related to changes)

**Comment by pdehaan** [comment]
This is super cool! 👍 
Not sure if we can randomize the experiment owner/feature/startdate a bit more in Storybook.

Also noticed the minor edge case of adding that sorting triangle can cause some columns to shift.

File: app/tests/integration/nimbus/test_e2e_create_experiment.py [hunk 1]
```diff
@@ -193,12 +193,12 @@ def test_create_new_experiment_remote_settings_reject(selenium, base_url):
     modal.decline_changes()
 
     # Load home page and wait for experiment to show in the Drafts tab
```
Comment (lmorchard) [inline app/tests/integration/nimbus/test_e2e_create_experiment.py:195]: This gets the failing integration test to pass, but it's kind of a weird fix. 

The thing that it was tripping on was `HomePage(selenium, base_url).wait_for_page_to_load()` - and as far as I can tell, that just looks for the presence of `.directory-table` in the DOM.

As far as I can tell, that element is in the DOM when it fails. And this same method is used in other tests and works. But, for some reason, it consistently fails in this test.

File: app/tests/integration/nimbus/test_e2e_create_experiment.py [hunk 2]
```diff
@@ -193,12 +193,12 @@ def test_create_new_experiment_remote_settings_reject(selenium, base_url):
...
-    selenium.get(base_url)
+    drafts_tab_url = f"{base_url}?tab=drafts"
+    selenium.get(drafts_tab_url)
     experiment_found = False
     for attempt in range(45):
         try:
-            home = HomePage(selenium, base_url).wait_for_page_to_load()
-            home.tabs[-1].click()
+            home = HomePage(selenium, drafts_tab_url)
```
Comment (lmorchard) [inline app/tests/integration/nimbus/test_e2e_create_experiment.py:201]: his gets the failing integration test to pass, but it's kind of a weird fix.

The thing that it was tripping on was HomePage(selenium, base_url).wait_for_page_to_load() - and as far as I can tell, that just looks for the presence of .directory-table in the DOM.

As far as I can tell, that element is in the DOM when it fails. And this same method is used in other tests and works. But, for some reason, it consistently fails in this test.

Comment (pdehaan) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:201]: /Paging Dr @jrbenny35

**Comment by lmorchard** [comment]
Alright, I think I've puttered with this enough. CI is passing, going to mark ready for review. Fresh stories:

* https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/1abcb6eb7047255614fd13cc1c1d2e10fa44819e/nimbus-ui/index.html?path=/story/pages-home--basic
* https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/1abcb6eb7047255614fd13cc1c1d2e10fa44819e/nimbus-ui/index.html?path=/story/pages-home-directorytable--basic

File: app/experimenter/nimbus-ui/src/lib/test-utils.tsx [hunk 3]
```diff
@@ -116,3 +117,16 @@ export const assertSerializerMessages = async (
...
   }
 };
+
+export const CurrentLocation = () => {
+  const location = useLocation();
+  return (
+    <p className="p-3">
+      Location:{" "}
+      <code data-testid="location">
+        {location.pathname}
```
Comment (jodyheavener) [inline app/experimenter/nimbus-ui/src/lib/test-utils.tsx:127]: Looks like this is always rendering `/slug/edit`

**Review by jodyheavener — APPROVED** [review]
Trying to find things but coming up pretty well empty, this just works really well

**Review by jaredlockhart — APPROVED** [review]
Looks and works great @lmorchard !  🎉 🎉 🎉 🎉

File: app/experimenter/nimbus-ui/src/lib/test-utils.tsx (hunk 3 above)
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/lib/test-utils.tsx:127]: Yeah, that's just the canned value from mocks. Threw it in there just to make sure the path didn't get altered by the query parameter manipulation

**Comment by lmorchard** [comment]
Gotta take the r+'s and merge to clear the way for issue #5958
